File: ui/src/apis/media.ts
import type { MediaUploadResponse } from "@/types";
import { API_ENDPOINTS } from "@/consts";
import { apiClient } from "@/utils/axios";

/**
 * Upload an avatar image through isme's backend proxy. The proxy forwards to
 * medioa2's public upload API with isme's server-side API key — the browser
 * never holds the key. Returns the stable, servable URL plus the medioa file id.
 */
export const uploadMedia = async (file: File): Promise<MediaUploadResponse["data"]> => {
	const formData = new FormData();
	formData.append("file", file);

	const response = await apiClient.post<MediaUploadResponse>(API_ENDPOINTS.MEDIA_UPLOAD, formData, {
		headers: { "Content-Type": "multipart/form-data" },
	});
	return response.data.data;
};

File: internal/domains/media/models/media.go
package models

import (
	"errors"
	"io"
)

// MaxAvatarSize caps an avatar upload at 2MB (the mock states "max 2MB"). The
// cap is enforced server-side; the UI also pre-checks so the user gets instant
// feedback before the round trip.
const MaxAvatarSize int64 = 2 * 1024 * 1024

// allowedAvatarContentTypes is the MIME allowlist for avatar uploads. Anything
// outside this set is rejected before the file reaches medioa.
var allowedAvatarContentTypes = map[string]bool{
	"image/png":  true,
	"image/jpeg": true,
	"image/webp": true,
}

// UploadRequest is the parsed multipart upload the isme backend proxies to
// medioa. File content is streamed, never buffered into the model. Size is the
// declared multipart size used for the 2MB cap check.
type UploadRequest struct {
	File        io.Reader
	FileName    string
	ContentType string
	Ext         string
	Size        int64
}

func (r UploadRequest) Validate() error {
	if r.File == nil {
		return errors.New("file is required")
	}
	if r.FileName == "" {
		return errors.New("file name is required")
	}
	if !allowedAvatarContentTypes[r.ContentType] {
		return errors.New("file must be a PNG, JPEG, or WebP image")
	}
	if r.Size > MaxAvatarSize {
		return errors.New("file is over 2MB")
	}
	return nil
}

// UploadResponse is what the isme UI receives after a successful proxied
// upload: a stable, browser-servable URL plus the medioa file id and metadata.
type UploadResponse struct {
	URL      string `json:"url"`
	FileID   string `json:"file_id"`
	FileName string `json:"file_name"`
	FileSize int64  `json:"file_size"`
}
